network/src/router.rs

```diff
@@ -174,18 +174,19 @@ impl<P: ProvideRuntimeApi + Send + Sync + 'static, E, N, T> Router<P, E, N, T> w
 
 				if let Some(work) = producer.map(|p| self.create_work(c_hash, p)) {
 					trace!(target: "validation", "driving statement work to completion");
-					let exit = self.fetcher.exit().clone().unit_error().compat();
-					let work = work.select2(exit).then(|_| Ok(()));
-					self.fetcher.executor().spawn(work);
+
+					let work = select(work, self.fetcher.exit().clone())
+						.map(drop);
+					let _ = self.fetcher.executor().spawn(work);
 				}
 			}
 		}
 	}
 
 	fn create_work<D>(&self, candidate_hash: Hash, producer: ParachainWork<D>)
-		-> impl Future<Item=(),Error=()> + Send + 'static
+		-> impl Future<Output=()> + Send + 'static
```

Comment:
Also this?

Reply:
This one can't be done due to lifetime issues, but I'll put in an async block so that the code is cleaner.